pytato/target/pycuda/__init__.py

```diff
@@ -0,0 +1,184 @@
+        :arg entrypoint: Name of the entrypoint
+        """
+        pass
+
+
+@dataclass(repr=False, eq=False)
+class BoundCUDAGraphProgram(BoundProgram):
+    """
+    A wrapper for executing python programs with bound arguments.
+
+    .. automethod:: __call__
+    .. automethod:: copy
+    .. automethod:: with_transformed_program
+    """
+    expected_arguments: FrozenSet[str]
+    entrypoint: str
+
+    @cached_property
+    def _compiled_function(self) -> Callable[..., Any]:
+        variables_after_execution: Dict[str, Any] = {
+            "_MODULE_SOURCE_CODE": self.program  # helps pudb
+        }
+        exec(self.program, variables_after_execution)
+        assert callable(variables_after_execution[self.entrypoint])
+        return variables_after_execution[  # type: ignore[no-any-return]
+            self.entrypoint]
+
+    @cached_property
+    def _bound_argment_names(self) -> Set[str]:
+        return set(self.bound_arguments.keys())
+
+    def __call__(self,  *args: Any, **kwargs: Any) -> Any:
+
+        if args:
+            raise ValueError(f"'{type(self).__call__}' does not take positional"
+                             " arguments.")
+
+        if set(kwargs.keys()) & self._bound_argment_names:
+            raise ValueError("Got arguments that were previously bound: "
+                    f"'{set(kwargs.keys()) & set(self.bound_arguments.keys())}'.")
+
+        updated_kwargs = dict(self.bound_arguments)
+        updated_kwargs.update(kwargs)
+        updated_kwargs = {kw: arg
+                          for kw, arg in updated_kwargs.items()
+                          if kw in self.expected_arguments}
+        return self._compiled_function(**updated_kwargs)
+
+    def copy(self, **kwargs: Any) -> BoundCUDAGraphProgram:
+        from dataclasses import replace
+        return replace(self, **kwargs)
+
+    def with_transformed_program(self, *args: Any, **kwargs: Any
+                                 ) -> BoundCUDAGraphProgram:
+        raise ValueError("Cannot transform python program.")
+
+# }}}
+
+
+# {{{ numpy-like python targets
+
+class NumpyLikePythonTarget(Target, ABC):
+
+    @abstractproperty
+    def numpy_like_module_name(self) -> str:
+        pass
+
+    @abstractproperty
+    def numpy_like_module_name_shorthand(self) -> str:
+        pass
+
+    @abstractmethod
+    def bind_program(self,
+                     program: str,
+                     entrypoint: str,
+                     expected_arguments: FrozenSet[str],
+                     bound_arguments: Mapping[str, Any]) -> BoundCUDAGraphProgram:
+        pass
+
+# }}}
+
+
+# {{{ cudagraph-numpy target
+
+
+class CUDAGraphTarget(NumpyLikePythonTarget):
+    """
+    A target that generates code for a python program by offloading array
+    operations to :mod:`cudagraph.cudagraph`.
+    """
+    
+    def __init__(self, allocator=None):
+        import pycuda
+        import pycuda.autoinit
```

Comment:
This should be lazier i.e. we should init a context only when we are executing an `ExecGraph`.